### ISSUE TITLE
AutoMapper upgrade to version 9.0.0 results in error: MissingMethodException IgnoreAllPropertiesWithAnInaccessibleSetter()

### ISSUE THREAD

**Issue by adria3a3** [JonPSmith/EfCore.GenericServices#33]
We are trying to update all packages in our solution to the latest versions. When upgrading AutoMapper to version 9.0.0 to following error occurs: 

```
System.MissingMethodException : Method not found: 'AutoMapper.IMappingExpression`2<System.__Canon,System.__Canon> AutoMapper.IMappingExpression`2.IgnoreAllPropertiesWithAnInaccessibleSetter()'.
   at GenericServices.Setup.Internal.CreateConfigGenerator.ConfigGenerator`2.AddSaveMappingToProfile(Profile writeProfile)
   at GenericServices.Setup.UnitTestSetup.AddSingleDto[TDto](SpecificUseData utData)
   at GenericServices.Setup.UnitTestSetup.SetupSingleDtoAndEntities[TDto](DbContext context, IGenericServicesConfig publicConfig)
```

The depedencies of GenericServices are:
AutoMapper (>= 7.0.1)
Microsoft.AspNetCore.JsonPatch (>= 2.1.1)
Microsoft.EntityFrameworkCore (>= 2.1.1)

It would be nice that GenericServices is compatible with the latest version of AutoMapper.

**Comment by JonPSmith** [comment]
Hmm. I didn't know that! I'm busy on a contract so I'm not sure when I will get to it, but its now on my list!

**Comment by Tealons** [comment]
We are also doing an upgrade of our application and are running into this problem. Do you have a ballpark ETA @JonPSmith ? We can also look into creating a PR if that's something that would be accepted by you.

**Comment by JonPSmith** [comment]
I plan to look at this during the upgrade into EF Core 3. Hopefully in October.

If you want to look at how to replace the missing `IgnoreAllPropertiesWithAnInaccessibleSetter` method then please do. You can PR if before October, or just send me what you have found.

**Comment by Tealons** [comment]
I'm not sure if there will be time in the coming weeks because one of our team members is leaving. Can give a ping in this issue when you start working on it? I will do the same if my team has the time to do it.

**Comment by baoduy** [comment]
Let me take a look on this hope I could able to give an PR in couple of days.

**Comment by JonPSmith** [comment]
Hi @baoduy,

Fine by me. Do make sure the unit tests are updated/passed. 

I will be working on the NET Core 3.0 version next week.

**Comment by JonPSmith** [comment]
Hi @baoduy and others,

I have just released a new version of EfCore.GenericServices, version 2.1.0-preview001 which uses AutoMapper 9.0.0. 

**Please can you check this works with your system soon**, as I plan to release a NetStandard2.0 and NetStandard2.1 version next week.

PS. Version 9 of AutoMapper has a `IgnoreAllPropertiesWithAnInaccessibleSetter` method so I can only assume signature changed from version 7 `IgnoreAllPropertiesWithAnInaccessibleSetter` didn't work with the version 9 AutoMapper.

**Comment by baoduy** [comment]
@JonPSmith 
Wow so fast, I still stuck with my system. However, I will test the package now and let you know if any issues as the current in-develop project is using your package.

**Comment by baoduy** [comment]
Hi @JonPSmith 
I think the package is fine. However, as Auto Mapper 9 is not supporting  'CreateMissingTypeMaps'. We need to fix some mapping issues. 

The below class of package 'EfCore.GenericBizRunner' needs to be fixed as well.
```cshap
internal class BizRunnerProfile : Profile
    {
        public BizRunnerProfile()
        {
            //CreateMissingTypeMaps = true;
        }
    }
```

**Comment by baoduy** [comment]
Seems upgrading to version 9 needs a huge effort to add `[AutoMap(typeof(SourceClass))]` to all Dto class even it already `ILinkToEntity<SourceClass>`.

**Comment by JonPSmith** [comment]
In reply to @baoduy,

Are you aware on how GenericServices sets up AutoMapper. It builds a separate read/write IMapper from any normal AutoMapper usage so that they don't get confused. It does this using the standard, i.e. not the new `[AutoMap(...)]` approach. See [this page in the docs](https://github.com/JonPSmith/EfCore.GenericServices/wiki/GenericServices-and-DTOs). 

This means GenericServices won't use the`[AutoMap(...)]` attribute when setting up the mappings. If you want to change the mappings then you use the `PerDtoConfig<TDto, TEntity>` class to alter the mapping. I don't plan to support the `[AutoMap(...)]` attribute as I couldn't implement the readonly feature on properties.

Can you explain why you need `CreateMissingTypeMaps = true;` in EfCore.GenericBizRunner? Do you need nested DTOs that you can't define?

**Comment by JonPSmith** [comment]
PS. GenericServices supports nested DTOs - see [Nested DTOs](https://github.com/JonPSmith/EfCore.GenericServices/wiki/GenericServices-and-DTOs#nested-dtos)

**Comment by Tealons** [comment]
I tested in one of our smaller project and it seems to work fine! We however now have issues with Automapper and xUnit we need to solve :)

Gives us some time to wait for the stable release of this change and the upgrade to .NET Core 3.0

**Comment by baoduy** [comment]
Hi @JonPSmith 
Sorry seems my message is confusing you. The `CreateMissingTypeMaps = true;` need to be removed from class `BizRunnerProfile` as this property deprecated on Auto Mapper 9.

**Comment by JonPSmith** [comment]
Ok, that makes more sense! Thanks.

**Comment by baoduy** [comment]
Hi @JonPSmith 
I have tested the latest version of EfCore.GenericServices with a PRD app. It is working fine still now. As there a many breaking on Auto Mapper v9 so I'm using the Auto Mapper v8 as of now.

**Comment by JonPSmith** [comment]
I have just released version 3.0.0 to support EF Core >=3.0 (and still support EF Core >=2.1). This uses AutoMapper 9.0.0.

Note: There is also a new `EfCore.GenericBizRunner` (Version 4.0.0) with EF Core 3.0 support and that also uses AutoMapper 9.0.0.
